src/abi/rand.rs

```diff
@@ -0,0 +1,24 @@
+use object::ProcessRef;
+use nabi::{Result, Error};
+use nebulet_derive::nebulet_abi;
+use arch::x64::rand::RdRand;
+use rand_core::RngCore;
+
+fn get_rdrand() -> Result<RdRand> {
+    RdRand::new().map_or(Err(Error::UNAVAILABLE), |v| Ok(v))
+}
+
+static mut RDRAND : Option<Result<RdRand>> = None;
+
+#[nebulet_abi]
+pub fn random_u32(_process: &ProcessRef) -> Result<u32> {
```

Comment:
I'm pretty sure `rdrand` is a very slow instruction. It'd be best to seed a prng when the `rand` device is initialized. Also, combine more than just the `rdrand` instruction. I'd recommend using `rand_core::JitterRng` as well.

Reply:
Actually, I looking into it more, and `rdrand` apparently has a very high throughput. My point still stands that it needs to be combined with other sources of randomness however.

Reply:
Then we get into the "how do we find entropy to seed a PRNG" debate.

I wanted to avoid that for my first pull request, by providing randomness which may be of lower quality first.

If we do have to solve that problem right now, I'll implement rdseed, which is designed to provide entropy, maybe jitter as well, and I think that's all the entropy sources we have access to without virtio support. I'll provide knobs to decide which sources may be enabled for entropy, but I don't believe any of them should be trusted by default.

Reply:
Also, while a kernel or service-side PRNG might be useful to prevent starving the entropy pool, for performance reasons the client should have its own PRNG. Therefore we don't have to care about IPC performance here.

Reply:
If we combine `rdrand`, `rdseed`, and jitter (all of which are available on that specific machine), I'll say it's safe.

Reply:
That's true, since this is just for drawing entropy for a CPRNG, let's stick to `rdseed` and `jitter` for generating a seed, since that's what they're for. Also, the abi interface should take a ptr and len and write random data into that buffer, instead of providing random numbers.

Reply:
Any of these can be hostile. They can also all fail, depending on hardware support, or on statistics that are collected in the case of test_timer and the JitterRng. So I'm fairly certain the API should be allowed to fail.

Reply:
Hmm, I'll have to think about this.

What's not clear to me is what the application should do if it can't get random data. Should it treat the system as compromised? At that point, does it matter if the random data is backdoored?